src/Network/DNS.hs

```diff
@@ -152,17 +152,22 @@ queryRaw (Class cls) (Name name) qtype = withCResState $ \stptr -> do
             resetErrno
             reslen <- c_res_query stptr dn (fromIntegral cls) qtypeVal resptr max_msg_size
 
-            unless (reslen <= max_msg_size) $
+            unless (reslen <= max_msg_size) $ do
+                c_res_nclose stptr
                 fail "res_query(3) message size overflow"
 
             errno <- getErrno
 
             when (reslen < 0) $ do
-                unless (errno == eOK) $
+                unless (errno == eOK) $ do
+                    c_res_nclose stptr
```

Comment:
Couldn't this call to `c_res_nclose` be placed directly after `getErrno`, rather than 3 times in each branch of the case distinction over `reslen` and `errno`?
If it does not affect `getErrno`, it could even be placed directly after `c_res_query`, could it?
(Same in the other occurrences below.)

Reply:
You are probably right, I'll test this.